Sources/GRPC/ConnectionManager.swift

```diff
@@ -282,51 +279,98 @@ internal class ConnectionManager {
+      guard case let .connecting(connecting) = self.state else {
+        self.invalidState()
+      }
+      multiplexer = connecting.readyChannelMuxPromise.futureResult
 
-      case let .ready(state):
-        channel = state.channel.eventLoop.makeSucceededFuture(state.channel)
+    case let .connecting(state):
+      multiplexer = state.readyChannelMuxPromise.futureResult
 
-      case let .transientFailure(state):
-        channel = state.readyChannelPromise.futureResult
+    case let .active(state):
+      multiplexer = state.readyChannelMuxPromise.futureResult
 
-      case let .shutdown(state):
-        channel = self.eventLoop.makeFailedFuture(state.reason)
-      }
+    case let .ready(state):
+      multiplexer = state.channel.eventLoop.makeSucceededFuture(state.multiplexer)
 
-      self.logger.debug("vending channel future", metadata: [
-        "connectivity_state": "\(self.state.label)",
-      ])
+    case let .transientFailure(state):
+      multiplexer = state.readyChannelMuxPromise.futureResult
 
-      return channel
+    case let .shutdown(state):
+      multiplexer = self.eventLoop.makeFailedFuture(state.reason)
     }
+
+    self.logger.debug("vending multiplexer future", metadata: [
+      "connectivity_state": "\(self.state.label)",
+    ])
+
+    return multiplexer
   }
 
-  /// Returns a future for the current channel, or future channel from the current connection
+  /// Returns a future for the current HTTP/2 stream multiplexer, or future HTTP/2 stream multiplexer from the current connection
   /// attempt, or if the state is 'idle' returns the future for the next connection attempt.
   ///
   /// Note: if the state is 'transientFailure' or 'shutdown' then a failed future will be returned.
-  internal func getOptimisticChannel() -> EventLoopFuture<Channel> {
+  private func getHTTP2MultiplexerOptimistic() -> EventLoopFuture<HTTP2StreamMultiplexer> {
+    func fulfillMuxPromiseGivenState(promise: EventLoopPromise<HTTP2StreamMultiplexer>) {
+      switch self.state {
+      case .idle:
+        // Unexpected to get here -
+        // we shouldn't be fulfilling if we've not yet worked out if we're succeeding or failing.
+        self.invalidState()
+      case .connecting:
+        // Unexpected to get here -
+        // we shouldn't be fulfilling if we've not yet worked out if we're succeeding or failing.
+        self.invalidState()
+      case let .active(state):
+        promise.succeed(state.multiplexer)
+      case let .ready(state):
+        promise.succeed(state.multiplexer)
+      case let .transientFailure(state):
+        // Provide the reason we failed transiently, if we can.
+        let error = state.reason ?? GRPCStatus(
+          code: .unavailable,
+          message: "Connection multiplexer requested while backing off"
+        )
+        promise.fail(error)
+      case let .shutdown(state):
+        promise.fail(state.reason)
+      }
+    }
+
     return self.eventLoop.flatSubmit {
```

Comment:
Since `getHTTP2Multiplexer` ensures we're on the event loop we no longer need to submit the work to it